sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoLifecycleTest.java

```diff
@@ -267,19 +290,17 @@ private void validate() {
                     "Function should have been torn down after exception",
                     value.finalState(),
                     is(CallState.TEARDOWN)));
-  }
 
-  @Test
-  @Category({ValidatesRunner.class, UsesStatefulParDo.class, UsesParDoLifecycle.class})
-  public void testTeardownCalledAfterExceptionInFinishBundleStateful() {
-    ExceptionThrowingFn fn = new ExceptionThrowingStatefulFn(MethodForException.FINISH_BUNDLE);
-    p.apply(Create.of(KV.of("a", 1), KV.of("b", 2), KV.of("a", 3))).apply(ParDo.of(fn));
-    try {
-      p.run();
-      fail("Pipeline should have failed with an exception");
-    } catch (Exception e) {
-      validate();
-    }
+    Arrays.stream(requiredCallStates)
+        .forEach(
+            callState ->
+                assertThat(
+                    "At least one "
+                        + callState
+                        + " lifecycle method should be called, got "
+                        + ExceptionThrowingFn.callStateMap.values(),
+                    ExceptionThrowingFn.callStateMap.values().stream()
+                        .anyMatch(value -> value.callStateVisited.containsKey(callState))));
```

Comment:
This only checks whether the required call state was reached, but isn't the order important?

Reply:
I believe the order is checked, because the terminal state is checked to be TEARDOWN, and the respective transitions are checked in  `ExceptionThrowinFn` `@Setup`, `@StartBundle`, `@ProcessElement`, etc.

Reply:
It would be definitely possible. On the other hand, this method is used only in with this DoFn (or `ExceptionThrowingStatefulFn` which performs the same checks), so it might be a little redundant. But I understand the concerns, so I will look into that. The main complication is that in some bundles the order might be different than in others (due to @ProcessElement missing). But it should not be that hard. :)